src/cmdline/mod.rs

```diff
@@ -343,17 +431,106 @@ impl Cmdline {
+    /// -> <boot args>
+    /// where <boot args> and <init args> can contain '--' only if double quoted and
+    /// <boot args> and <init args> contain at least one non-whitespace char each.
+    ///
+    /// Providing a str not following these rules might end up in undefined behaviour of
+    /// the resulting `Cmdline`.
+    ///
+    /// # Arguments
+    ///
+    /// * `cmdline_raw` - Contains boot params and init params of the cmdline.
+    /// * `capacity` - Capacity of the cmdline.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use linux_loader::cmdline::*;
+    /// let cl = Cmdline::try_from("foo -- bar", 100).unwrap();
+    /// assert_eq!(
+    ///     cl.as_cstring().unwrap().as_bytes_with_nul(),
+    ///     b"foo -- bar\0"
+    /// );
+    /// ```
+    pub fn try_from(cmdline_raw: &str, capacity: usize) -> Result<Cmdline> {
+        // The cmdline_raw argument should contain no more than one INIT_ARGS_SEPARATOR sequence
+        // that is not double quoted; in case the INIT_ARGS_SEPARATOR is found all chars following
+        // it will be parsed as init args.
+
+        if capacity == 0 {
+            return Err(Error::InvalidCapacity);
+        }
+
+        // Step 1: Extract boot args and init args from input by searching for INIT_ARGS_SEPARATOR.
+
+        // Check first occurrence of the INIT_ARGS_SEPARATOR that is not between double quotes.
+        // All chars following the INIT_ARGS_SEPARATOR will be parsed as init args.
+        let (mut boot_args, mut init_args) = match cmdline_raw
+            .match_indices(INIT_ARGS_SEPARATOR)
+            .find(|&separator_occurrence| {
+                Self::check_outside_double_quotes(&cmdline_raw[..(separator_occurrence.0)])
+            }) {
+            None => (cmdline_raw, ""),
+            Some((delimiter_index, _)) => (
+                &cmdline_raw[..delimiter_index],
+                // This does not overflow as long as `delimiter_index + INIT_ARGS_SEPARATOR.len()`
+                // is pointing to the first char after the INIT_ARGS_SEPARATOR which always exists;
+                // as a result, `delimiter_index + INIT_ARGS_SEPARATOR.len()` is less or equal to the
+                // length of the initial string.
+                &cmdline_raw[(delimiter_index + INIT_ARGS_SEPARATOR.len())..],
+            ),
+        };
+
+        boot_args = boot_args.trim();
+        init_args = init_args.trim();
+
+        // Step 2: Check if capacity provided for the cmdline is not exceeded and create a new `Cmdline`
+        // if size check passes.
+        let mut cmdline_size = boot_args.len().checked_add(1).ok_or(Error::TooLarge)?;
+
+        if !init_args.is_empty() {
+            cmdline_size = cmdline_size
+                .checked_add(INIT_ARGS_SEPARATOR.len())
+                .ok_or(Error::TooLarge)?;
+
+            cmdline_size = cmdline_size
+                .checked_add(init_args.len())
+                .ok_or(Error::TooLarge)?;
```

Comment:
nit: these `checked_add`s can be chained into something like

```rust
    let cmdline_size = boot_args
            .len()
            .checked_add(1)
            .and_then(|s| {
                if !init_args.is_empty() {
                    s.checked_add(INIT_ARGS_SEPARATOR.len() + init_args.len())
                } else {
                    Some(s)
                }
            })
            .ok_or(Error::TooLarge)?;
```

Reply:
As discussed, I'll let as it is, as long as we still need to have checked_add() for `INIT_ARGS_SEPARATOR.len() + init_args.len()`